### PR TITLE
[NeuralNet] Fix testset run

### PR DESCRIPTION
**Pull request by zhoonit**
- [NeuralNet] Fix testset run

```
This patch fixes testset running while cleaning up the code to reuse
repeating logics with callback

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: nntrainer/models/neuralnet.h [hunk 1]
```diff
@@ -63,10 +63,12 @@ using DatasetModeType = ml::train::DatasetModeType;
  * @brief     Statistics from running or training a model
  */
 typedef struct RunStats_ {
-  float accuracy; /** accuracy of the model */
-  float loss;     /** loss of the model */
+  float accuracy;     /** accuracy of the model */
+  float loss;         /** loss of the model */
+  int count;          /** number of iterations done on this stat */
```
Comment (kparichay) [inline nntrainer/models/neuralnet.h:68]: rename to `num_iterations`?

File: nntrainer/models/neuralnet.h [hunk 2]
```diff
@@ -63,10 +63,12 @@ using DatasetModeType = ml::train::DatasetModeType;
...
+  unsigned int right; /** number of right sample on this run */
```
Comment (kparichay) [inline nntrainer/models/neuralnet.h:69]: rename to `num_correct_predictions`? `right` is a bit vague.

File: nntrainer/models/neuralnet.cpp [hunk 3]
```diff
@@ -632,25 +632,26 @@ int NeuralNetwork::train_run() {
...
+   * @brief run a single epoch with given callback, @a auto is used instead of
+   * std::function for performance measure
+   * @param buffer buffer to run
+   * @param shuffle whether to shuffle or not
+   * @param on_iteration_fetch function that will recieve reference to stat,
+   * buffer which will be called every time data is fetched and set
+   * @param on_epoch_end function that will recieve reference to stat,
+   * buffer which will be called on the epoch end
+   */
+  auto run_epoch = [this, &in, &label, &in_dims, &label_dims](
```
Comment (kparichay) [inline nntrainer/models/neuralnet.cpp:645]: 💯

Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:645]: you can consider making it a private function itself than a lambda as there will be eventually many hooks supported for this.

Comment (zhoonit) [reply nntrainer/models/neuralnet.cpp:645]: I agree, But i'll delay for now :)

File: nntrainer/models/neuralnet.h (hunk 1 above)
Comment (zhoonit) [reply nntrainer/models/neuralnet.h]: Done

File: nntrainer/models/neuralnet.h (hunk 2 above)
Comment (zhoonit) [reply nntrainer/models/neuralnet.h]: Done

File: nntrainer/models/neuralnet.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:645]: IMO, the function will most likely be a static function which will additionally take `NetworkGraph` object, and return stats as now, so can remain in .cpp.

Comment (zhoonit) [reply nntrainer/models/neuralnet.cpp:645]: If it remains in .cpp with static, I guess this can stay here until we have the need.

**Review by kparichay — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**